File: backend/src/entities/alarm.py
from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, Field, model_validator

from entities.basis import Adresse, Eintrag, kennung as _kennung
from entities.planung import DIENSTSTELLE, Ort, Person, Planung, Tag

__all__ = ["Adresse", "Eintrag", "Karte", "Fahrzeug", "Einsatzmittelgruppe", "HinweisText",
           "HinweisCode", "Hinweis", "Alarm", "Fahrzeugvorlage", "Stichwortvorlage", "Kataloge",
           "Arbeitsmappe", "Planung", "Ort", "Person", "Tag", "Materialvorlage",
           "DIENSTSTELLE"]


class Karte(BaseModel):
    kab: str = ""
    fwPlan: str = ""
    ePlan: str = ""
    polarKoordinaten: str = ""


class Fahrzeug(Eintrag):
    """The catalogue entry this stands for; empty where the Funkrufname was typed by hand."""
    vorlageId: str = ""
    funkrufname: str = ""
    ezp: str = ""
    status: str = ""
    staerke: str = ""
    trupp: str = ""
    hinweis: str = ""
    """The sheet is addressed to one vehicle, and that vehicle's Funkrufname is the only cell
    printed on grey. At most one vehicle per Alarm carries this."""
    alarmFuer: bool = False


class Einsatzmittelgruppe(Eintrag):
    """One block of the Einsatzmittelaufgebot: a group name and its vehicles. The HA line above
    it prints the Alarm's Einsatzadresse, so the group carries no address of its own."""

    gruppe: str = ""
    fahrzeuge: list[Fahrzeug] = []


class HinweisText(Eintrag):
    typ: Literal["text"] = "text"
    text: str = ""


class HinweisCode(Eintrag):
    """
    A dispatch code and the path taken to reach it. The answers print as the numbered sentences
    that follow the code on the sheet, which is what the interrogation dialog writes.
    """

    typ: Literal["code"] = "code"
    code: str = ""
    meldung: str = ""
    antworten: list[str] = []


Hinweis = Annotated[Union[HinweisText, HinweisCode], Field(discriminator="typ")]


class Alarm(BaseModel):
    id: str
    behoerde: str = "Berliner Feuerwehr"
    titel: str = "Alarm für"
    einsatzNr: str = ""
    einsatzDatum: str = ""
    einsatzZeit: str = ""
    meldungDatum: str = ""
    meldungZeit: str = ""
    aPlatz: str = ""
    polizei: str = ""
    sonderrechte: str = ""
    arbeitsgruppe: str = ""
    wachalarmNr: str = ""
    """The catalogue entry this stands for; empty where the Stichwort was typed by hand."""
    stichwortId: str = ""
    stichwort: str = ""
    kurzinfo: str = ""
    anfahrtsadresse: Adresse = Adresse()
    einsatzadresse: Adresse = Adresse()
    karte: Karte = Karte()
    meldungsquelle: str = ""
    rueckrufnummer: str = ""
    anrufer: str = ""
    betroffener: str = ""
    meldender: str = ""
    wasIstPassiert: str = ""
    hinweise: list[Hinweis] = []
    einsatzmittel: list[Einsatzmittelgruppe] = []


class Fahrzeugvorlage(BaseModel):
    """
    A vehicle as the station keeps it: crew strength, Einsatzpunkt and status are known in
    advance and come along when the vehicle is picked. Each stays editable on the Alarm.

    An Alarm points at the id, so renaming the vehicle reaches every sheet that calls for it.
    """

    id: str = Field(default_factory=_kennung)
    funkrufname: str = ""
    staerke: str = ""
    ezp: str = ""
    status: str = ""
    """Wie viele Köpfe hineinpassen. Nur für den Ablaufplan; der Alarmzettel kennt das nicht."""
    plaetze: str = ""
    """Welche Fahrerlaubnisklasse es verlangt, gegen `Person.fahrerlaubnis` geprüft."""
    fuehrerschein: str = ""


class Materialvorlage(BaseModel):
    """
    Ein Stück Material, wie die Wache es führt. Ein Schritt zeigt darauf, damit ein Umbenennen
    jeden Plan erreicht, der es dabeihat.
    """

    id: str = Field(default_factory=_kennung)
    name: str = ""
    """
    Wie viel davon überhaupt da ist. Null heißt nicht erfasst — dann wird auch nicht gezählt,
    ob der Plan mehr verplant, als es gibt.
    """
    bestand: int = 0


class Stichwortvorlage(BaseModel):
    """A Stichwort is only its text, so the id is the whole reason an Alarm can follow a rename."""

    id: str = Field(default_factory=_kennung)
    text: str = ""


class Kataloge(BaseModel):
    """Suggestion lists the user maintains; the open data covers only the medical codes."""

    stichwoerter: list[Stichwortvorlage] = []
    fahrzeuge: list[Fahrzeugvorlage] = []
    status: list[str] = []
    trupp: list[str] = []
    """
    Die angelegten Orte. Ein Ort gehört der Wache und nicht einem einzelnen Übungstag — deshalb
    steht er hier neben den Fahrzeugen und nicht im Plan.
    """
    orte: list[Ort] = []
    """Was die Wache an Material führt. Neu Geschriebenes kommt von selbst dazu."""
    material: list[Materialvorlage] = []
    """
    Die Tage der Veranstaltung, das Personal und die Wörter, aus denen sich seine Rollen und
    Fahrerlaubnisklassen zusammensetzen. Alles, was die Wache dauerhaft führt, steht hier —
    damit es an einem Ort steht und nicht an zweien.
    """
    tage: list[Tag] = []
    personen: list[Person] = []
    rollen: list[str] = []
    fahrerlaubnisse: list[str] = []
    """One fixed value for the whole working set; every new Alarm starts with it."""
    arbeitsgruppe: str = ""
    """The station the sheets are written for; the Polar-Koordinaten are measured from it."""
    wache: Adresse = Adresse()
    """Wie die Dienststelle im Plan heißt. Sie ist ein Ort, den es immer gibt."""
    wacheName: str = ""
    """
    Wie viele Alarme die Leitstelle an einem Tag zählt. Daraus wird die Einsatznummer einer
    geplanten Lage: der Anteil des Tages, der bis zur Einsatzzeit vergangen ist. Null heißt,
    dass keine abgeleitet wird und die getippte stehen bleibt.
    """
    alarmeProTag: int = 2200

    def alle_orte(self) -> list[Ort]:
        """
        Die Dienststelle zuerst, dann die angelegten. Sie ist kein Eintrag, den jemand anlegt,
        sondern einer, den es gibt, solange die Wache eine Adresse hat — und deshalb auch keiner,
        den man löschen kann.
        """
        dienststelle = Ort(id=DIENSTSTELLE, sortierung=-1,
                           name=self.wacheName or "Dienststelle", adresse=self.wache)
        return [dienststelle, *self.orte]


class Arbeitsmappe(BaseModel):
    """The whole working set, as it is downloaded, uploaded and held in the browser."""

    @model_validator(mode="before")
    @classmethod
    def _orte_in_den_katalog(cls, werte: Any) -> Any:
        """
        Orte standen einmal im Plan. Sie gehören der Wache und nicht dem einzelnen Übungstag,
        also ziehen sie einmalig in den Katalog um — die Schritte zeigen ohnehin nur auf ihre id.
        """
        if not isinstance(werte, dict):
            return werte
        plan = werte.get("planung")
        kataloge = werte.get("kataloge")
        if not isinstance(plan, dict) or not isinstance(kataloge, (dict, type(None))):
            return werte
        alte = plan.get("orte")
        kataloge = kataloge or {}
        if alte and not kataloge.get("orte"):
            werte = {**werte, "kataloge": {**kataloge, "orte": alte}}
        return werte

    version: int = 1
    alarme: list[Alarm] = []
    kataloge: Kataloge = Kataloge()
    """Der Ablaufplan. Abschaltbar; solange er aus ist, ändert er am Alarmzettel nichts."""
    planung: Planung = Planung()

File: frontend/src/interfaces/Alarm.ts
/**
 * Anything that sits in a list carries an id and a sort key. Two people editing one Alarm touch
 * different entries far more often than the same one, so the merge works per entry — which needs
 * an identity that survives a reorder. The sort key is fractional: an entry moved between two
 * others takes the value between theirs, so moving one thing never rewrites the rest.
 */
export interface Eintrag {
    id: string
    sortierung: number
}

export interface Adresse {
    strasse: string
    hnr: string
    objekt: string
    plz: string
    ort: string
}

export interface Karte {
    kab: string
    fwPlan: string
    ePlan: string
    polarKoordinaten: string
}

export interface Fahrzeug extends Eintrag {
    /** The catalogue entry this stands for, if any. Empty where the Funkrufname was typed. */
    vorlageId: string
    funkrufname: string
    ezp: string
    status: string
    staerke: string
    trupp: string
    hinweis: string
    /** The vehicle this sheet is addressed to — printed on grey, and only ever one per Alarm. */
    alarmFuer: boolean
}

export interface Einsatzmittelgruppe extends Eintrag {
    gruppe: string
    fahrzeuge: Fahrzeug[]
}

export interface HinweisText extends Eintrag {
    typ: 'text'
    text: string
}

/** A dispatch code together with the path taken to reach it, as printed after the code. */
export interface HinweisCode extends Eintrag {
    typ: 'code'
    code: string
    meldung: string
    antworten: string[]
}

export type Hinweis = HinweisText | HinweisCode

export interface Alarm extends Eintrag {
    behoerde: string
    titel: string
    einsatzNr: string
    einsatzDatum: string
    einsatzZeit: string
    meldungDatum: string
    meldungZeit: string
    aPlatz: string
    polizei: string
    sonderrechte: string
    arbeitsgruppe: string
    wachalarmNr: string
    /** The catalogue entry this stands for, if any. Empty where the Stichwort was typed. */
    stichwortId: string
    stichwort: string
    kurzinfo: string
    anfahrtsadresse: Adresse
    einsatzadresse: Adresse
    karte: Karte
    meldungsquelle: string
    rueckrufnummer: string
    anrufer: string
    betroffener: string
    meldender: string
    wasIstPassiert: string
    hinweise: Hinweis[]
    einsatzmittel: Einsatzmittelgruppe[]
}

/** A vehicle as the station keeps it: picking it brings its strength, EZP and status along. */
export interface Fahrzeugvorlage {
    /** What an Alarm points at, so renaming the vehicle reaches every sheet that uses it. */
    id: string
    funkrufname: string
    staerke: string
    ezp: string
    status: string
    /** Wie viele Köpfe hineinpassen. Nur für den Ablaufplan; der Alarmzettel kennt das nicht. */
    plaetze: string
    /** Welche Fahrerlaubnisklasse es verlangt, gegen `Person.fahrerlaubnis` geprüft. */
    fuehrerschein: string
}

/** A Stichwort is only its text, so the id is the whole reason an Alarm can follow a rename. */
export interface Stichwortvorlage {
    id: string
    text: string
}

export interface Kataloge {
    stichwoerter: Stichwortvorlage[]
    fahrzeuge: Fahrzeugvorlage[]
    status: string[]
    trupp: string[]
    /** One fixed value for the whole working set; every new Alarm starts with it. */
    arbeitsgruppe: string
    /** The station the sheets are written for. The Polar-Koordinaten are measured from it. */
    wache: Adresse
}

export interface Arbeitsmappe {
    version: number
    alarme: Alarm[]
    kataloge: Kataloge
    /** Der Ablaufplan. Abschaltbar und ohne Wirkung auf den Alarmzettel, solange er aus ist. */
    planung: Planung
}

import {aPlatzKennung, festnetznummer, mobilnummer, zufallsname} from '../scripts/generator'
import {leerePlanung, type Planung} from './Planung'

export const ARBEITSMAPPE_VERSION = 1

export function leereAdresse(): Adresse {
    return {strasse: '', hnr: '', objekt: '', plz: '', ort: ''}
}

/** New entries go to the end; `naechste` reads the highest sort key already in the list. */
export function naechste(eintraege: Eintrag[]): number {
    return eintraege.reduce((groesste, eintrag) => Math.max(groesste, eintrag.sortierung), -1) + 1
}

export function leeresFahrzeug(sortierung = 0): Fahrzeug {
    return {
        id: crypto.randomUUID(), sortierung, vorlageId: '',
        funkrufname: '', ezp: '', status: '', staerke: '', trupp: '', hinweis: '', alarmFuer: false,
    }
}

export function leereGruppe(sortierung = 0): Einsatzmittelgruppe {
    return {
        id: crypto.randomUUID(), sortierung,
        gruppe: 'keine Gruppe',
        fahrzeuge: [leeresFahrzeug()],
    }
}

export function leererHinweisText(sortierung = 0): HinweisText {
    return {id: crypto.randomUUID(), sortierung, typ: 'text', text: ''}
}

export function leererHinweisCode(sortierung = 0): HinweisCode {
    return {id: crypto.randomUUID(), sortierung, typ: 'code', code: '', meldung: '', antworten: []}
}

/** The Stichwörter the Berlin AAO uses, as the starting point for a fresh working set. */
export const STICHWOERTER = [
    'BRAND 1', 'BRAND 2', 'BRAND 3', 'BRAND 4', 'BRAND 6', 'BRAND 8', 'BRAND 10',
    'BRAND K.', 'BRAND M.',
    'TH 1', 'TH 2', 'TH 3', 'TH K.', 'TH M.',
]

export function stichwortvorlagen(): Stichwortvorlage[] {
    return STICHWOERTER.map(text => ({id: crypto.randomUUID(), text}))
}

export function leererAlarm(): Alarm {
    const heute = new Date()
    const datum = heute.toLocaleDateString('de-DE')
    const zeit = heute.toLocaleTimeString('de-DE', {hour: '2-digit', minute: '2-digit'})
    return {
        id: crypto.randomUUID(),
        sortierung: 0,
        behoerde: 'Berliner Feuerwehr',
        titel: 'Alarm für',
        einsatzNr: '',
        einsatzDatum: datum,
        einsatzZeit: zeit,
        meldungDatum: datum,
        meldungZeit: zeit,
        aPlatz: aPlatzKennung(),
        stichwortId: '',
        polizei: 'N',
        sonderrechte: 'J',
        arbeitsgruppe: '',
        wachalarmNr: '',
        stichwort: '',
        kurzinfo: '',
        anfahrtsadresse: leereAdresse(),
        einsatzadresse: leereAdresse(),
        karte: {kab: '', fwPlan: '', ePlan: '', polarKoordinaten: ''},
        meldungsquelle: festnetznummer(),
        rueckrufnummer: mobilnummer(),
        anrufer: zufallsname(),
        betroffener: '',
        meldender: '',
        wasIstPassiert: '',
        hinweise: [],
        einsatzmittel: [leereGruppe()],
    }
}

export function leereArbeitsmappe(): Arbeitsmappe {
    return {
        version: ARBEITSMAPPE_VERSION,
        alarme: [],
        planung: leerePlanung(),
        kataloge: {
            stichwoerter: stichwortvorlagen(), fahrzeuge: [], status: [], trupp: [],
            arbeitsgruppe: '', wache: leereAdresse(),
        },
    }
}
